### PR TITLE
TypeScript support

### PR DESCRIPTION
**Pull request by amireh**
Things to do/check:

[ts-loader](https://github.com/TypeStrong/ts-loader) support:
- [x] compiles files
- [x] reports module-level errors
- [x] reports program-level errors (aggregated)
- [x] works in `--watch` mode
- [x] reports errors in changed module in `--watch` mode
- [x] works with happy's cache (caching of code/map)
- [ ] works with the `transpileOnly` compiler option
- [ ] works with `--bail`
- [ ] replays errors/warnings for cached modules

[awesome-typescript-loader](https://github.com/s-panferov/awesome-typescript-loader) support:
- [x] compiles files
- [ ] reports module-level errors
- [ ] reports program-level errors (aggregated, this sorta works but can contain duplicate errors)
- [ ] works in `--watch` mode
- [ ] reports errors in changed module in `--watch` mode
- [ ] works with happy's cache (caching of code/map)
- [ ] works with `--bail`
- [ ] replays errors/warnings for cached modules

To try out these changes, use the `happypack-support` branches in my forks of those repositories. For example, to install the happypack-supporting version of `ts-loader`:

``` js
// package.json
{
  "devDependencies": {
    "ts-loader": "amireh/ts-loader#happypack-support"
  }
}
```

You also need to clone this PR branch.

### CODE REVIEW

**Comment by amireh** [comment]
One thing I noticed is that "warming up" a typescript loader takes a considerable amount of time (~700ms in my samples). In the regular webpack build, using either loader, it took around ~800ms to build the example, which contains only 3 sources. With happypack, there's about ~100ms overhead for the entire build, so it takes 900ms **if you're using a single thread**, and ~700ms for every consecutive thread. It's very unfortunate. Compiling the example using happypack in 2 threads takes somewhere between 1500-1700ms (using either loader - it's _almost_ the same.)

I digged a bit into `tsserver` trying to see if we can make the loaders use/connect to a stand-alone version of the servers, but my research was futile. There seems to be very little documentation around `tsserver`.

In my opinion, this should either be solved at the TypeScript level (e.g. have `tsserver` distribute the compilation to multiple threads internally, like what HappyPack does to Webpack), or find out how to launch stand-alone instances of `tsserver` that one can have running in the background, then instruct our loaders to connect to them.

This *should* be doable because I checked Sublime Text's TS plugin and it seems to have a tsserver instance in the background to do the diagnostics/type-checking.

**TL;DR** - don't expect fascinating results unless your builds take ~15 seconds or longer already.

**Comment by amireh** [comment]
It might be useful if you guys can pull in some TypeScript experts to this thread because this is out of my scope.

**Comment by TheLarkInn** [comment]
CC @jbrantly @s-panferov @danielrosenwasser is this 'warmup' that @amireh a typescript issue?

**Comment by amireh** [comment]
Also, is there any large open-source TypeScript project that can be built with webpack that we (I) can test on?

**Comment by TheLarkInn** [comment]
angular2-webpack-starter is a good one.

**Comment by IAMtheIAM** [comment]
I could not get either to work.

Your fork of awesome-typescript-loader produced:

```
Happy[ts]: Version: 2.1.1. Using cache? no. Threads: 5 (shared pool)
Happy[ts]: All set; signalling webpack to proceed.
  0% compile 10% 0/1 build modules 70% 1/1 build modules 40% 1/2 build modules 70% 2/2 build modules 50% 2/3 build modules 70% 3/3 build modules 54% 3/4 build modules 46% 3/5 build modules 40% 3/6 build modules 35% 3/7 build modules 32% 3/8 build modules 40% 4/8 build modules 36% 4/9 build modules 34% 4/10 build modules 31% 4/11 build modules 37% 5/11 build modules 35% 5/12 build modules 40% 6/12 build modules 45% 7/12 build modules 42% 7/13 build modules 40% 7/14 build modules 44% 8/14 build modules 42% 8/15 build modules 46% 9/15 build modules 50% 10/15 build modules 47% 10/16 build modules 45% 10/17 build modules 43% 10/18 build modules 41% 10/19 build modules 40% 10/20 build modules 38% 10/21 build modules 37% 10/22 build modules 36% 10/23 build modules 35% 10/24 build modules 34% 10/25 build modules 33% 10/26 build modules 32% 10/27 build modules 31% 10/28 build modules 30% 10/29 build modules 30% 10/30 build modules 29% 10/31 build modules 28% 10/32 build modules 28% 10/33 build modules 27% 10/34 build modules 27% 10/35 build modules 26% 10/36 build modules 26% 10/37 build modules 25% 10/38 build modules 25% 10/39 build modules 25% 10/40 build modules 24% 10/41 build modules 24% 10/42 build modules 23% 10/43 build modules[TypeError: compiler.plugin is not a function] 'TypeError: compiler.plugin is not a function\n    at setupWatchRun (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:231:14)\n    at ensureInstance (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:197:5)\n    at _callee3$ (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:98:65)\n    at tryCatch (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:62:40)\n    at GeneratorFunctionPrototype.invoke [as _invoke] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:336:22)\n    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:95:21)\n    at C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:48:65\n    at new Promise (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\core-js\\modules\\es6.promise.js:193:7)\n    at __awaiter (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:28:12)\n    at compiler (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:85:12)'
 23% 10/44 build modules 25% 11/44 build modules[TypeError: compiler.plugin is not a function] 'TypeError: compiler.plugin is not a function\n    at setupWatchRun (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:231:14)\n    at ensureInstance (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:197:5)\n    at _callee3$ (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:98:65)\n    at tryCatch (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:62:40)\n    at GeneratorFunctionPrototype.invoke [as _invoke] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:336:22)\n    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:95:21)\n    at C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:48:65\n    at new Promise (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\core-js\\modules\\es6.promise.js:193:7)\n    at __awaiter (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:28:12)\n    at compiler (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:85:12)'
 26% 12/44 build modules 27% 13/44 build modules 27% 13/45 build modules 26% 13/46 build modules 26% 13/47 build modules 27% 14/47 build modules 29% 15/47 build modules 30% 16/47 build modules 31% 17/47 build modules 31% 17/48 build modules 30% 17/49 build modules 30% 17/50 build modules 30% 17/51 build modules 29% 17/52 build modules 29% 17/53 build modules 28% 17/54 build modules 30% 18/54 build modules 31% 19/54 build modules 32% 20/54 build modules 33% 21/54 build modules 32% 21/55 build modules 32% 21/56 build modules 32% 21/57 build modules 31% 21/58 build modules 31% 21/59 build modules 32% 22/59 build modules 33% 23/59 build modules 34% 24/59 build modules 35% 25/59 build modules 36% 26/59 build modules 36% 26/60 build modules 37% 27/60 build modules 38% 28/60 build modules 39% 29/60 build modules 40% 30/60 build modules 41% 31/60 build modules 40% 31/61 build modules 41% 32/61 build modules 42% 33/61 build modules 43% 34/61 build modules 44% 35/61 build modules 45% 36/61 build modules 44% 36/62 build modules 45% 37/62 build modules 46% 38/62 build modules 47% 39/62 build modules 48% 40/62 build modules 48% 40/63 build modules 47% 40/64 build modules 46% 40/65 build modules 46% 40/66 build modules 45% 40/67 build modules 45% 40/68 build modules 46% 41/68 build modules 47% 42/68 build modules 47% 43/68 build modules 48% 44/68 build modules 49% 45/68 build modules 49% 45/69 build modules 50% 46/69 build modules 50% 47/69 build modules 51% 48/69 build modules 52% 49/69 build modules 53% 50/69 build modules 52% 50/70 build modules 52% 50/71 build modules 51% 50/72 build modules 52% 51/72 build modules 53% 52/72 build modules 54% 53/72 build modules 54% 54/72 build modules 54% 54/73 build modules 53% 54/74 build modules 53% 54/75 build modules 52% 54/76 build modules 53% 55/76 build modules 54% 56/76 build modules 53% 56/77 build modules 53% 56/78 build modules 52% 56/79 build modules 53% 57/79 build modules 54% 58/79 build modules 54% 59/79 build modules 55% 60/79 build modules 56% 61/79 build modules 57% 62/79 build modules 57% 63/79 build modules 58% 64/79 build modules 59% 65/79 build modules 60% 66/79 build modules 60% 67/79 build modules 61% 68/79 build modules 62% 69/79 build modules 63% 70/79 build modules 63% 71/79 build modules 64% 72/79 build modules 65% 73/79 build modules 66% 74/79 build modules 66% 75/79 build modules 67% 76/79 build modules 67% 76/80 build modules 67% 77/80 build modules    3830ms build modules
 71% seal 5ms seal
 73% optimize[TypeError: compiler.plugin is not a function] 'TypeError: compiler.plugin is not a function\n    at setupWatchRun (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:231:14)\n    at ensureInstance (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\instance.js:197:5)\n    at _callee3$ (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:98:65)\n    at tryCatch (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:62:40)\n    at GeneratorFunctionPrototype.invoke [as _invoke] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:336:22)\n    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\node_modules\\babel-polyfill\\node_modules\\regenerator-runtime\\runtime.js:95:21)\n    at C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:48:65\n    at new Promise (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\core-js\\modules\\es6.promise.js:193:7)\n    at __awaiter (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:28:12)\n    at compiler (C:\\Source\\POC\\Atlas.Prototype\\src\\starterproject\\node_modules\\awesome-typescript-loader\\dist.babel\\index.js:85:12)'
19ms optimize
 75% hashing 4ms hashing
 76% create chunk assets19ms create chunk assets
 78% additional chunk assets 4ms additional chunk assets
 80% optimize chunk assets362ms optimize chunk assets
 90% optimize assets141ms optimize assets
 95% emit22ms emit
Version: webpack 1.13.1

ERROR in ./src/polyfills.ts
Module parse failed: C:\Source\POC\Atlas.Prototype\src\starterproject\node_modules\happypack\loader.js?id=ts!C:\Source\POC\Atlas.Prototype\src\starterproject\node_modules\awesome-typescript-loader\dist.babel\entry.js!C:\Source\POC\Atlas.Prototype\src\starterproject\src\polyfills.ts Unexpected token (1:8)
You may need an appropriate loader to handle this file type.
SyntaxError: Unexpected token (1:8)
```

**Comment by amireh** [comment]
:scream: that is not exciting news. I didn't get the chance to check out the angular project yet so I can't verify, but will take a look as soon as I can.

Also, more work was done on `ts-loader` than `awesome-typescript-loader` so far so there's definitely something up. Is your project open-source by any chance or would the starter project do just as well?

**Comment by DanielRosenwasser** [comment]
>  is this 'warmup' that @amireh mentions a typescript issue?

This might just be the initial parse/bind time (parsing & binding `lib.d.ts` is on its own often a bottleneck) - you can see if `--skipLibCheck` from our nightly builds helps there at all.

**Comment by TheLarkInn** [comment]
@danielrosenwasser is this a tsconfig option that can be used?

**Comment by DanielRosenwasser** [comment]
It should be - check out https://github.com/Microsoft/TypeScript/pull/8735

**Comment by IAMtheIAM** [comment]
@amireh I will post my project source code up at my next opportunity and post the link here.

**Comment by IAMtheIAM** [comment]
Hey @amireh, here is the source code for the project I'm working on, and trying to get HappyPack working with typescript.

https://drive.google.com/open?id=0B6I99RX-ZH6gcEhKMGZrNGFuZDA

To run, follow these commands

```
npm run install:all
npm run start
```
Install will take about 5 minutes usually. To view the app, go to

```[http://localhost:4000](http://localhost:4000)```

The only view of the Angular2 app you'll have access to is the login screen, but that should be fine since you're working on compilation stuff and not the app components.

**Comment by IAMtheIAM** [comment]
Hey @amireh , just wondering if this is still on your mental to do list as a possibility?

**Comment by amireh** [comment]
@IAMtheIAM 

> just wondering if this is still on your mental to do list as a possibility?

I've asked webpack core for help on this, to see how webpack may be able to help typescript loaders do their job without the use of internal objects, see [this discussion](https://github.com/webpack/webpack/issues/2975) if you're interested. That's what's making it very difficult to support them in happypack.

I didn't do additional work on this PR because I'm not sure the approach is feasible at all; we might end up with worse performance than without happypack like you've seen so far or it will make things a maintenance nightmare on my end. I'm hoping that I could get some clues out of that discussion which would enable us to apply fixes in happypack ahead of time until webpack core officially supports this use case (this won't be until webpack 3 I believe so it's still some time away.)

So, yes, it's still on the to-do list and I hope it gets moving soon. :)

**Comment by abergs** [comment]
@amireh Any new updates on this? Our build times using webpack is about 120s... we would love to use Happypack to speed things up.

**Comment by amireh** [comment]
@abergs I'd love to help out in this but given how complicated ts loaders are, and the fact that it's a hit/miss thing (i don't really know if it _can_ be solved at happypack's layer at all) it's going to take a lot of time to figure this out, and I don't have that time. If you're willing to sponsor the work, maybe I can investigate but otherwise I'm afraid there isn't much to do.

Did you guys try the suggestion of using the ts loader only for transpiling without the error reporting? I've heard that that works with happypack (and then you can use your editor or a separate linter to receive program-wide reports.) That may be a viable alternative?

**Comment by abergs** [comment]
Hey @amireh - Thanks for getting back to me. I'm now indeed using ts-loader only for transpiling but have yet to try HappyPack.
